{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import helper\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tifs = helper.read_all_tifs(positive_strand_only=True, return_dict=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "genome = helper.organize_genome_by_chrom(helper.read_bedfile(), positive_strand_only=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "densities = pickle.load(open(\"densities.pickle\", \"rb\"))\n",
    "for chrom in helper.kYeastChroms:\n",
    "    densities[chrom] = densities[chrom][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "\n",
    "save_to_pgf=True\n",
    "\n",
    "np.random.seed(1135)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "if save_to_pgf: \n",
    "    matplotlib.use(\"pgf\")\n",
    "    matplotlib.rcParams.update({\n",
    "        \"pgf.texsystem\": \"pdflatex\",\n",
    "        'font.family': 'serif',\n",
    "        'text.usetex': True,\n",
    "        'pgf.rcfonts': False,\n",
    "    })\n",
    "else:\n",
    "    matplotlib.rcdefaults()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:helper:Reading bedfile...\n",
      "INFO:helper:Done reading bedfile ../indexes/yeast-all.bed\n",
      "INFO:helper:Processed TIFs in chromosome 1...\n",
      "INFO:helper:Processed TIFs in chromosome 2...\n",
      "INFO:helper:Processed TIFs in chromosome 3...\n",
      "INFO:helper:Processed TIFs in chromosome 4...\n",
      "INFO:helper:Processed TIFs in chromosome 5...\n",
      "INFO:helper:Processed TIFs in chromosome 6...\n",
      "INFO:helper:Processed TIFs in chromosome 7...\n",
      "INFO:helper:Processed TIFs in chromosome 8...\n",
      "INFO:helper:Processed TIFs in chromosome 9...\n",
      "INFO:helper:Processed TIFs in chromosome 10...\n",
      "INFO:helper:Processed TIFs in chromosome 11...\n",
      "INFO:helper:Processed TIFs in chromosome 12...\n",
      "INFO:helper:Processed TIFs in chromosome 13...\n",
      "INFO:helper:Processed TIFs in chromosome 14...\n",
      "INFO:helper:Processed TIFs in chromosome 15...\n",
      "INFO:helper:Processed TIFs in chromosome 16...\n",
      "INFO:helper:Done reading TIF file... returning a dictionary\n"
     ]
    }
   ],
   "source": [
    "from metagene_helper import select_one_random_junction, kJunctionCandidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:helper:Done reading TIF file... returning a pandas\n"
     ]
    }
   ],
   "source": [
    "unsorted_tifs = helper.read_all_tifs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_all_junctions_chrom(chrom, tifs):\n",
    "    splits = []\n",
    "    for gene, gtifs in tifs[chrom].items():\n",
    "        junctions = helper.generate_5utr_isoform_starts(gtifs)\n",
    "        splits = splits + junctions[1:]\n",
    "    return splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 44\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(143657, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 44\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(33419, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 1184\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(143602, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 1184\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(183721, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 171\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(236353, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 171\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(270212, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5884\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(236325, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5884\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(549350, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 51\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(111897, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 51\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(70110, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 1798\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(111285, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 1798\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(111526, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 342\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(447984, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 342\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(139478, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 13103\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(447875, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 13103\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(861840, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 136\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(148130, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 136\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(302640, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5265\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(146962, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5265\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(131624, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 49\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(147093, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 49\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(246036, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 1650\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(146994, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 1650\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(95880, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 254\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(494443, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 254\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(81262, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 8309\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(494396, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 8309\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(438287, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 135\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(104223, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 135\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(489491, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 4290\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(104178, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 4290\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(493863, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 89\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(353875, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 89\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(177104, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 2625\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(353636, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 2625\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(338867, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 163\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(435124, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 163\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(639818, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5367\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(434915, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5367\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(116788, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 156\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(437745, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 156\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(402146, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5328\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(437300, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5328\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(571485, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 212\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(147834, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 212\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:helper:(340150, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 6974\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(147811, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 6974\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(853393, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 215\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(267083, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 215\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(110995, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 8075\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(266923, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 8075\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(306579, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 193\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(627431, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 193\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(676125, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5981\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(627376, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 5981\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(526066, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 254\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(325140, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 254\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(1058282, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 8174\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(324805, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 8174\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(1056548, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 231\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(554589, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 231\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(610439, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 7790\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(554549, '+')\n",
      "INFO:helper:Generating metagene plot... Number of overlapping sites is: 7790\n",
      "INFO:helper:<class 'tuple'>\n",
      "INFO:helper:type of splits[0] is <class 'tuple'>\n",
      "INFO:helper:(468751, '+')\n"
     ]
    }
   ],
   "source": [
    "big_meta = [0] * (2 * n)\n",
    "fake_meta = [0] * (2 * n)\n",
    "\n",
    "bad_meta = [0] * (2 * n)\n",
    "bad_random_meta = [0] * (2 * n)\n",
    "\n",
    "\n",
    "np.random.seed(1135)\n",
    "\n",
    "for chrom in helper.kYeastChroms:\n",
    "    splits = select_one_random_junction(chrom, tifs)\n",
    "    meta = helper.make_metagene_plot((densities[chrom], []),\n",
    "                                     splits,\n",
    "                                     \"\",\n",
    "                                     n,\n",
    "                                     n,\n",
    "                                     plot=False)\n",
    "    big_meta = np.add(big_meta, meta)\n",
    "\n",
    "    fake_junctions = [(x, \"+\") for x in np.random.choice(\n",
    "        kJunctionCandidates[chrom],\n",
    "        size=len(splits),\n",
    "        replace=False)]\n",
    "    meta = helper.make_metagene_plot((densities[chrom], []),\n",
    "                                     fake_junctions,\n",
    "                                     \"\",\n",
    "                                     n,\n",
    "                                     n,\n",
    "                                     plot=False)\n",
    "    fake_meta = np.add(fake_meta, meta)\n",
    "    \n",
    "    all_junctions = get_all_junctions_chrom(chrom, tifs)\n",
    "    meta = helper.make_metagene_plot((densities[chrom], []),\n",
    "                                     all_junctions,\n",
    "                                     \"\",\n",
    "                                     n,\n",
    "                                     n,\n",
    "                                     plot=False)\n",
    "    bad_meta = np.add(bad_meta, meta)\n",
    "    \n",
    "    fake_junctions = [(x, \"+\") for x in np.random.choice(\n",
    "        kJunctionCandidates[chrom],\n",
    "        size=len(all_junctions),\n",
    "        replace=False)]\n",
    "    meta = helper.make_metagene_plot((densities[chrom], []),\n",
    "                                     fake_junctions,\n",
    "                                     \"\",\n",
    "                                     n,\n",
    "                                     n,\n",
    "                                     plot=False)\n",
    "    bad_random_meta = np.add(bad_random_meta, meta)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(range(-n, n), big_meta)    \n",
    "plt.title(\"Metagene plot of non-overlaping ioform positions\")\n",
    "plt.xlabel(\"Distance from junction (nt)\")\n",
    "plt.ylabel(\"Aggregated Read density\")\n",
    "plt.savefig(\"final_plots/good_metagene.pgf\")\n",
    "\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(range(-n, n), fake_meta)    \n",
    "plt.title(\"Metagene plot of non-overlaping random positions\")\n",
    "plt.xlabel(\"Distance from junction (nt)\")\n",
    "plt.ylabel(\"Aggregated Read density\")\n",
    "plt.savefig(\"final_plots/random_metagene.pgf\")\n",
    "\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(range(-50, 50), bad_meta[50:150])    \n",
    "plt.title(\"Metagene plot of all ioform positions\")\n",
    "plt.xlabel(\"Distance from junction (nt)\")\n",
    "plt.ylabel(\"Aggregated Read density\")\n",
    "plt.savefig(\"final_plots/all_metagene.pgf\")\n",
    "\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(range(-50, 50), bad_random_meta[50:150])    \n",
    "plt.title(\"Metagene plot of random genomic positions\")\n",
    "plt.xlabel(\"Distance from junction (nt)\")\n",
    "plt.ylabel(\"Aggregated Read density\")\n",
    "plt.savefig(\"final_plots/random_big_metagene.pgf\")\n",
    "\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
